data/src/main/java/com/microsoft/azure/kusto/data/auth/CloudInfo.java

```diff
@@ -0,0 +1,132 @@
+
+    private final boolean loginMfaRequired;
+    private final String loginEndpoint;
+    private final String kustoClientAppId;
+    private final String kustoClientRedirectUri;
+    private final String kustoServiceResourceId;
+    private final String firstPartyAuthorityUrl;
+
+    public CloudInfo(boolean loginMfaRequired, String loginEndpoint, String kustoClientAppId, String kustoClientRedirectUri, String kustoServiceResourceId, String firstPartyAuthorityUrl) {
+        this.loginMfaRequired = loginMfaRequired;
+        this.loginEndpoint = loginEndpoint;
+        this.kustoClientAppId = kustoClientAppId;
+        this.kustoClientRedirectUri = kustoClientRedirectUri;
+        this.kustoServiceResourceId = kustoServiceResourceId;
+        this.firstPartyAuthorityUrl = firstPartyAuthorityUrl;
+    }
+
+    public static void manuallyAddToCache(String clusterUrl, CloudInfo cloudInfo){
+        synchronized (cache) {
+            cache.put(clusterUrl, cloudInfo);
+        }
+    }
+
+    public static CloudInfo retrieveCloudInfoForCluster(String clusterUrl) throws DataServiceException {
+        synchronized (cache) {
+            CloudInfo cloudInfo = cache.get(clusterUrl);
+            if (cloudInfo != null) {
+                return cloudInfo;
+            }
+
+            CloudInfo result;
+
+            try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
+                HttpGet request = new HttpGet(clusterUrl + "/" + METADATA_ENDPOINT);
+                request.addHeader(HttpHeaders.ACCEPT_ENCODING, "gzip,deflate");
+                request.addHeader(HttpHeaders.ACCEPT, "application/json");
+                try (CloseableHttpResponse response = httpClient.execute(request)) {
+                    int statusCode = response.getStatusLine().getStatusCode();
+                    if (statusCode == 200) {
+                        String content = EntityUtils.toString(response.getEntity());
+                        if (content == null || content.equals("") || content.equals("{}")) {
+                            throw new DataServiceException(clusterUrl, "Error in metadata endpoint, received no data", true);
+                        }
+                        result = parseCloudInfo(content);
+                    } else if (statusCode == 404) {
+                        result = DEFAULT_CLOUD;
+                    } else {
+                        String errorFromResponse = EntityUtils.toString(response.getEntity());
+                        throw new DataServiceException(clusterUrl,
+                                "Error in metadata endpoint, got code: " + statusCode + "\nWith error: " + errorFromResponse, true);
+                    }
+                }
+            } catch (IOException ex) {
+                throw new DataServiceException(clusterUrl, "IOError when trying to retrieve CloudInfo", ex, true);
+            }
+            cache.put(clusterUrl, result);
```

Comment:
Does this require sync?

Reply:
Yes - 
1. HashMaps aren't really necessarily thread safe, so protecting `put` calls is a good idea
2. Since all of the function is synced, this means that only one thread will actually make the web reqeust

Reply:
I read the scope wrong again.
Sorry